File: lang/csharp/src/vtfpp/DistanceMapping.cs
using System;

namespace sourcepp.vtfpp;

public static class DistanceMapping
{
	public enum Dither
	{
		NONE = 0,
		GRADIENT_TANGENT,
	}

	public enum Flags : uint
	{
		NONE           = 0,
		DISTANCEAA     = 1 << 0,
		EUCLIDEAN      = 1 << 1,
		SAMPLECENTERED = 1 << 2,
	}

	public static byte[] AlphaToDistance(ReadOnlySpan<byte> buffer, ImageFormat inFormat, ImageFormat outFormat, ushort width, ushort height, ushort resizeX, ushort resizeY, bool srgb, ref bool valveQuirks, float distanceSpread = 1.0f, float alphaThreshold = 0.04f, Flags flags = Flags.NONE, Dither dither = Dither.NONE, ImageConversion.ResizeFilter filter = ImageConversion.ResizeFilter.NICE, ImageConversion.ResizeEdge edge = ImageConversion.ResizeEdge.CLAMP)
	{
		var valveQuirksInt = Convert.ToInt32(valveQuirks);
		unsafe
		{
			var valveQuirksIntPtr = &valveQuirksInt;
			var output = new sourcepp.Buffer(DLL.vtfpp_distance_mapping_alpha_to_distance_ex(buffer, (ulong)buffer.Length, inFormat, outFormat, width, height, resizeX, resizeY, Convert.ToInt32(srgb), distanceSpread, alphaThreshold, flags, dither, filter, edge, valveQuirks ? (nint)valveQuirksIntPtr : 0)).Read<byte>();
			valveQuirks = Convert.ToBoolean(valveQuirksInt);
			return output;
		}
	}
}

File: lang/c/src/vtfppc/DistanceMapping.cpp
#include <vtfppc/DistanceMapping.h>

#include <sourceppc/Helpers.h>

using namespace sourceppc;
using namespace vtfpp;

VTFPP_API sourcepp_buffer_t vtfpp_distance_mapping_alpha_to_distance(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e inFormat, vtfpp_image_format_e outFormat, uint16_t width, uint16_t height, uint16_t reduceX, uint16_t reduceY, int srgb) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);

	return convert::toBuffer(DistanceMapping::alphaToDistance({reinterpret_cast<const std::byte*>(buffer), bufferLen}, convert::cast(inFormat), convert::cast(outFormat), width, height, reduceX, reduceY, srgb));
}

VTFPP_API sourcepp_buffer_t vtfpp_distance_mapping_alpha_to_distance_ex(const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e inFormat, vtfpp_image_format_e outFormat, uint16_t width, uint16_t height, uint16_t reduceX, uint16_t reduceY, int srgb, float distanceSpread, float alphaThreshold, vtfpp_distance_mapping_flags_e flags, vtfpp_distance_mapping_dither_e dither, vtfpp_image_conversion_resize_filter_e filter, vtfpp_image_conversion_resize_edge_e edge, int* valveQuirks) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, SOURCEPP_BUFFER_INVALID);

	if (!valveQuirks) {
		return convert::toBuffer(DistanceMapping::alphaToDistance({reinterpret_cast<const std::byte*>(buffer), bufferLen}, convert::cast(inFormat), convert::cast(outFormat), width, height, reduceX, reduceY, srgb, distanceSpread, alphaThreshold, convert::cast(flags), convert::cast(dither), convert::cast(filter), convert::cast(edge), nullptr));
	}
	bool valveQuirksBool = *valveQuirks;
	const auto out = convert::toBuffer(DistanceMapping::alphaToDistance({reinterpret_cast<const std::byte*>(buffer), bufferLen}, convert::cast(inFormat), convert::cast(outFormat), width, height, reduceX, reduceY, srgb, distanceSpread, alphaThreshold, convert::cast(flags), convert::cast(dither), convert::cast(filter), convert::cast(edge), &valveQuirksBool));
	*valveQuirks = valveQuirksBool;
	return out;
}
